### ISSUE TITLE
Implement Symbolic Reasoning Engine with AST Operations

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#695]
## Goal
Enable Klareco to perform **multi-hop logical reasoning** by combining deterministic symbolic inference over ASTs with minimal learned components (retrieval, reranking).

## Motivation
Traditional LLMs rely on emergent reasoning from pretraining. Klareco should use **explicit symbolic rules** over structured AST representations, making reasoning:
- **Explainable**: Full provenance via AST Trail
- **Reliable**: Logic rules are deterministic
- **Efficient**: ~50M params vs 70B+ for comparable reasoning

### SymbolicReasoner
```python
class SymbolicReasoner:
    """Perform logical inference over ASTs"""
    
    def __init__(self):
        self.retriever = ASTRetriever()           # Already exists
        self.fact_extractor = FactExtractor()     # Partially exists
        self.inference_engine = InferenceEngine()  # NEW
        self.ast_trail = ASTTrail()               # NEW (explainability)
    
    def reason(self, question_ast):
        # 1. Extract structured query from question AST
        query = self.extract_query(question_ast)
        # "Kiu fondis Esperanton?" → query(who, founded, Esperanto)
        
        # 2. Retrieve relevant fact ASTs
        fact_asts = self.retriever.retrieve(query)
        self.ast_trail.add_step('retrieve', query, fact_asts, source='corpus')
        
        # 3. Extract structured facts from ASTs
        facts = [self.fact_extractor.extract(ast) for ast in fact_asts]
        
        # 4. Apply symbolic reasoning rules
        answer = self.inference_engine.infer(query, facts)
        self.ast_trail.add_step('infer', facts, answer, source='rule:transitivity')
        
        # 5. Generate answer AST
        answer_ast = self.construct_answer_ast(answer)
        
        return answer_ast, self.ast_trail
```

### InferenceEngine (100% Deterministic)
```python
class InferenceEngine:
    """Apply first-order logic rules to AST-derived facts"""
    
    RULES = [
        # Transitivity: founded(X, Y) ∧ created(Y, Z) → contributed_to(X, Z)
        lambda facts: self.transitivity(facts),
        
        # Temporal ordering: happened(E1, T1) ∧ happened(E2, T2) ∧ T1 < T2 → before(E1, E2)
        lambda facts: self.temporal_order(facts),
        
        # Property inheritance: instance_of(X, Y) ∧ has_property(Y, P) → has_property(X, P)
        lambda facts: self.inheritance(facts),
        
        # Spatial reasoning: in(X, Y) ∧ in(Y, Z) → in(X, Z)
        lambda facts: self.spatial_containment(facts),
        
        # Negation: ¬fact(X) ∧ query(X) → answer(no, reason)
        lambda facts: self.negation(facts),
    ]
    
    def infer(self, query, facts):
        # Apply all rules to derive new facts
        derived_facts = set(facts)
        
        # Fixed-point iteration
        while True:
            new_facts = set()
            for rule in self.RULES:
                new_facts.update(rule(derived_facts))
            
            if new_facts.issubset(derived_facts):
                break  # Fixed point reached
            
            derived_facts.update(new_facts)
        
        # Find fact matching query
        return self.match_query(query, derived_facts)
```

### ASTTrail (Explainability)
```python
class ASTTrail:
    """Track reasoning steps as AST transformations"""
    
    def __init__(self):
        self.steps = []
    
    def add_step(self, operation, input_ast, output_ast, source):
        self.steps.append({
            'operation': operation,  # 'retrieve', 'infer', 'combine'
            'input': input_ast,
            'output': output_ast,
            'source': source,  # Citation or rule name
        })
    
    def explain(self):
        """Generate human-readable explanation"""
        explanation = []
        for i, step in enumerate(self.steps, 1):
            explanation.append(f"Step {i} ({step['operation']}): {deparse(step['output'])}")
            if step['source']:
                explanation.append(f"  Source: {step['source']}")
        return "\n".join(explanation)
```

### Question
"Kiu fondis la unuan internacian lingvon?"
(Who founded the first international language?)

### Reasoning Steps
```
Step 1 (retrieve): Find facts about international languages
  - Zamenhof founded Esperanto (1887)
  - Esperanto is an international language
  - Volapük was founded by Schleyer (1879)
  - Volapük is an international language

Step 2 (infer - temporal ordering): 1879 < 1887 → Volapük before Esperanto
  Source: rule:temporal_order

Step 3 (infer - first): min(year) → Volapük is first international language
  Source: rule:temporal_extrema

Step 4 (infer - transitivity): Schleyer founded Volapük ∧ Volapük is first → Schleyer founded first
  Source: rule:transitivity

Answer: "Schleyer fondis la unuan internacian lingvon."

AST Trail:
  1. Retrieved 4 facts from corpus
  2. Applied temporal ordering rule
  3. Applied extrema rule (first = min)
  4. Applied transitivity rule
```

### Phase 1: Fact Extraction Enhancement
- [ ] Extend fact extractor to handle temporal facts
- [ ] Extract spatial relations (in, on, near)
- [ ] Extract causal relations (caused, led to)
- [ ] Test on corpus: coverage >80%

### Phase 2: Inference Engine (Deterministic)
- [ ] Implement 5-10 core inference rules
- [ ] Fixed-point iteration algorithm
- [ ] Query matching logic
- [ ] Unit tests for each rule

### Phase 3: AST Trail System
- [ ] Implement step tracking
- [ ] Explainability formatting
- [ ] CLI: `klareco reason --explain`
- [ ] Visualization (optional)

### Phase 4: Integration
- [ ] Combine retriever + inference engine
- [ ] Add to orchestrator/instruction handler
- [ ] Evaluation on multi-hop QA benchmarks
- [ ] Compare to baseline (no reasoning)

## Success Criteria
- [ ] Multi-hop reasoning accuracy >70% on test set
- [ ] AST Trail provides full provenance for all answers
- [ ] Inference rules are deterministic and testable
- [ ] Integration with RAG pipeline (#680, #690)
- [ ] CLI: `klareco reason "Kiu..." --explain`
- [ ] Documentation: Supported reasoning types

## Dependencies
- ASTRetriever (already exists)
- FactExtractor (partially exists, needs extension)
- M1 Selectional Preference (#687) - filters implausible facts
- Reranker (#686) - ranks fact relevance
- Semantic ontology (#654) - provides schema for reasoning

## Related Issues
- #687: M1 Selectional Preference (filters facts before reasoning)
- #686: Reranker (ranks retrieved facts)
- #680: Extractive answer generation (uses reasoning)
- #690: Abstractive answer generation (uses reasoning)
- #654: Semantic ontology (provides reasoning schema)
- #661: RST discourse relations (for summarization reasoning)

## Training Pipeline Position
```
Question → Parser → Query AST → Retriever → Fact ASTs
                                    ↓
                            M1 Filter (#687)
                                    ↓
                            Reranker (#686)
                                    ↓
                    [Inference Engine] (deterministic)
                                    ↓
                            Answer AST + Trail
```

## Inference Rules (Initial Set)

1. **Transitivity**: A→B ∧ B→C ⟹ A→C
2. **Temporal Ordering**: time(E1) < time(E2) ⟹ before(E1, E2)
3. **Temporal Extrema**: min(time) ⟹ first, max(time) ⟹ last
4. **Property Inheritance**: instance_of(X, Y) ∧ has(Y, P) ⟹ has(X, P)
5. **Spatial Containment**: in(X, Y) ∧ in(Y, Z) ⟹ in(X, Z)
6. **Negation**: ¬fact(X) ∧ query(X) ⟹ answer(no)
7. **Causality**: caused(A, B) ∧ caused(B, C) ⟹ led_to(A, C)
8. **Set Membership**: element(X, S) ∧ property(S, P) ⟹ property(X, P)
9. **Comparison**: greater(A, B) ∧ greater(B, C) ⟹ greater(A, C)
10. **Counting**: count(S) ∧ query(how_many) ⟹ answer(count(S))

### Create Esperanto Multi-Hop QA Dataset
- Temporal reasoning: "Kiu estis la unua...?"
- Spatial reasoning: "Kio estas en la urbo en Pollando?"
- Causal reasoning: "Kio kaŭzis...?"
- Comparison: "Kiu estas pli granda ol...?"
- Counting: "Kiom da...?"

Target: 500 questions with gold answers + reasoning chains

## Future Extensions
- Probabilistic reasoning (confidence scores)
- Abductive reasoning (best explanation)
- Analogical reasoning (pattern matching)
- Counterfactual reasoning (what-if scenarios)
- Integration with semantic ontology for richer reasoning

## Key Advantage
Traditional LLM reasoning is a **black box**. Klareco reasoning is **fully explainable** via AST Trail, showing exactly which facts were retrieved, which rules were applied, and why.

**Comment by marctjones** [comment]
Closing as part of a 2026-09-08 backlog cleanup (see PLAN.md, DESIGN.md, CLAUDE.md's merge gate).

This issue predates the merge gate (adopted 2026-07-13) and the AST-native orchestrator rewrite. It describes work against the pre-orchestrator RAG pipeline or the learned model stack (M1/M2 embeddings, compositional embeddings, entity classifiers, neural rerankers) that was pruned from the repo -- klareco/embeddings/, klareco/models/, and klareco/summarization/ no longer exist. It has no milestone and was never benchmarked, matching exactly the failure mode CLAUDE.md documents (25 capability issues filed 2026-05-26, 8 implemented, none benchmarked).

Labeled archived-2026-09 so this batch is recoverable: gh issue list --state closed --label archived-2026-09. If this capability is still wanted, re-file it against the current architecture naming the benchmark metric and test the merge gate requires.
